pkg/gadgets/interface.go

```diff
@@ -108,33 +104,47 @@ type EventEnricherSetter interface {
 	SetEventEnricher(func(ev any) error)
 }
 
-type StartStopGadget interface {
-	Start() error
-	Stop()
-}
-
-// RunGadget is an alternative to StartStopGadget that expects the result to be emitted using the EventHandler
-// as well but doesn't need to be stopped (manually)
+// RunGadget is an interface for gadgets that are run in the background and emit
+// events as soon as they occur.
 type RunGadget interface {
-	Run() error
-}
+	// Run is expected to run the gadget and emits the events using the
+	// EventHandler. It is expected to be blocking and return only when the
+	// context is done, after which the gadget should clean up all resources.
+	// Notice that Run() will be called after operators are installed.
+	Run(GadgetContext) error
+}
+
+// RunWithResultGadget is an alternative to RunGadget that returns the result
+// of the gadget only at the end of its execution.
+type RunWithResultGadget interface {
+	// RunWithResult follows the same rules as Run() but instead of using an
+	// EventHandler to emit the events, it returns the result of the gadget as a
+	// byte array after the context is done.
+	RunWithResult(GadgetContext) ([]byte, error)
+}
+
+// InitRunClose is a gadget interface that must be initialized before the
+// operators are installed. This is useful when the gadget needs to be kept
+// up-to-date with the containers that need to be traced, as the operators will
+// start sending notifications of this as soon as it is installed.
+type InitRunClose interface {
+	// Init is expected to initialize the gadget. It will be called before
+	// Run() and before the operators are installed.
+	Init(GadgetContext) error
 
-// StartStopAltGadget is an alternative interface to StartStop, as some gadgets
-// already have the Start() and Stop() functions defined, but with a different signature
-// After we've migrated to the interfaces, the gadgets should be altered to use the
-// Start/Stop signatures of the above StartStopGadget interface.
-type StartStopAltGadget interface {
-	StartAlt() error
-	StopAlt()
-}
+	// Run, in this case, is expected to just wait for the context to be done
+	// and emit the events using the EventHandler. And, also also in this case,
+	// it is also expected to be blocking and return only when the context is
+	// done.
+	Run(GadgetContext) error
 
-type CloseGadget interface {
+	// Close is expected to clean up all the resources allocated by the
+	// gadget. It will be called after Run() returns and after the operators
+	// have been uninstalled.
 	Close()
```

Comment:
Maybe add `GadgetContext` as parameter here as well? It's nothing expensive or confusing, but it might help not storing things, if we can get them inside `Close()` as well.

Reply:
I don't have a good example right now (maybe params only needed in `Close()` or checking the embedded context) and in general, I agree that the interfaces should be as clean and simple as possible. However, we're defining an API here that we hopefully don't have to change often - having the GadgetContext also for `Close()` doesn't hurt IMHO as it doesn't introduce anything new and is merely a pointer - but it may come in handy someday (maybe even after we extend the context functionality itself).

Reply:
Make sense. Done.

Reply:
I couldn't add it because there are some gadgets that needs to implement `Close(gadgetContext)` but they already implement `Close()` for the Tracer CR. I added a TODO on the interface and we can manage it on #1396.